### DISCUSSION TITLE
6장

### DISCUSSION THREAD

**Discussion by JoungMinJu** [somefood/cs-study#65] [공부만이 살길]
## 1. SQL의 기초적인 기술 규칙
1. SQL문의 마지막에 딜리미터를 붙인다.
2. 키워드의 대문자와 소문자는 구별하지 않는다.
3. 정수는 기존 그대로 쓴다. 문자열이나 날짜 시각은 작은 따옴표로 감싼다. 

## 2. 테이블을 조작하는 SQL 종류 요약

1. SELECT =  어디서 (FROM  테이블명) 무엇을 가지고 온다(SELECT 컬럼명)
2. WHERE 조건 (조건에는 비교 연산자 AND와 OR이 자주 사용됨)
3. DISTINCT 활용한 중복 제거 가능
4. ORDER 로 정렬
5. GROUP BY로 구분
6. HAVING으로 나눈 그룹에 조건 추가
7. UPDATE로 데이터 갱신 (UPDATE 테이블명 SET 컬럼명=값)
8. INSERT로 데이터 삽입
9. DELETE로 데이터 삭제

추가적으로 더 정리하고 싶은 부분을 정리한 내용은 아래와 같음.

### 2-1. 테이블을 요약하는 함수.
함수엔 크게 두 가지의 종류가 있다.

1. 복수 행에 대해 집계를 수행하는 함수(=집약함수)
2. 단일 해의 값에 대해 조작이나 계산을 수행하는 함수

**집약함수**의 대표적인 다섯 가지는 아래와 같다.

1. COUNT = 테이블의 행수를 알려줌
2. SUM =  테이블의 수치 데이터를 합계
3. AVG = 테이블의 수치 데이터 평균을 구함
4. MAX = 테이블의 임의열 데이터 중 최대 값을 구함
5. MIN = 테이블의 임의열 데이터 중 최소 값을 구함

이러한 집야감수는 기본적으로 **NULL을 제외**하고 집계한다.
하지만 **COUNT(*)**는 NULL을 포함한 전체 행을 집계한다.

또한 SUM, AVG를 제외한 나머지 함수는 수치 데이터 외에도 사용 수 있다.
예를 들어, city 컬럼의 MAX는 'Y'로 시작하는 'Yosu', MIN은 'A'로 시작하는 'Andong'이 된다.

### 2-2. 문자열을 집약하는 GROUP_CONCAT
SQL 표준에는 없지만 MySQL에는 GROUP_CONCAT 함수가 있다. 
GROUP_CONCAT은 문자열에 대한 집계를 **문자열의 결합**으로 수행한다.
근데 이 경우에는 예상 외의 길이를 가진 결과가 return 되는 경우가 많기 떄문에, GROUP_CONCAT 함수의 결과를 저장하거나 표시하는 경우 주의할 필요가 있다. 

> 시스템 변수로 상한선(1024바이트)이 정해져 있어서 그 이상의 문자열은 잘리기 때문

### 2-3. 데이터를 변경하는 UPDATE문

데이터 변경시엔 `UPDATE  테이블명 SET 컬럼명=값` 사용.
하지만 얘는 SET에 명시한 것 외에는 변경되지 않기 떄문에 주로 `UPDATE 테이블명 SET 컬럼명 = 값 WHERE 조건` 사용한다.

**복수 열의 동시 갱신**도 가능하다.

`UPDATE CITY SET NAME = '시흥', POPULATION = 432390 WHERE COUNTRYCODE = "KOR" AND DISTRICT = "Kyonggi" AND NAME = "Sihung";`

### 2-4. 데이터를 입력하는 INSERT문

행 단위로 수행되기 때문에 테이블 정의를 명확하게 알아야 한다.

`SHOW CREATE TABLE 테이블명\G` 로 확인 가능.

단순히 열의 정보를 알고 싶은 경우에는 Oracle과 호환되는 명령인 `desc 테이블명`으로도 확인 가능하다.

INSERT문의 기본 구분은 `INSERT INTO 테이블명(열1[,열2,..]) VALUES(값1[, 값2,...])` 이다.

이때 테이블의 모든 컬럼에 대해 값을 설정하면 열 리스트를 생략해도 된다.

`VALUES DEFAULT`로 해당 행에 해당하는 기본 값을 입력할 수도 있다. 아니면 컬럼 리스트에서 해당 값을 제거하면 자동으로 기본 값이 들어간다.

`CREATE TABLE 테이블명1 LIKE 테이블명2`란 구문으로 테이블명2와 같은 구조의 테이블을 작성할 수도 있다. 

INSERT문 하나로 신규 복수 행을 입력할 수도 있다.
`INSERT INTO city(name, countrycode, district, poluation) VALUES ('Gimpo', '~', '~','~'), ('~','~','~','~')...`

## 3. 뷰를 작성하고 복수 테이블에서 선택하기

**뷰를 사용하는 이점**은 무엇일까

1.  복잡한 SELECT 문을 일일이 매번 기술할 필요가 없다.
2. 필요한 열과 행만 사용자에게 보여줄 수 있다.

### 3-1. 뷰를 생성
`CREATE VIEW 뷰명 (컬럼명1[,컬럼명2, ...]) AS SELECT 문;`

### 3-2. 뷰로의 입력 및 갱신 제한

기본적으로는 어떤 행이 대응하는지 모르거나 어떤 값을 넣으면 좋을지 모르는 경우에는 갱신할 수 없다. 

예를 들어 "GROUP BY로 집약한 수치"나 "DISTINCT"로 얻은 값을 갱신하는 경우엔 어떤 수치를 갱신하는 것이 좋을지 모르기 때문에 불가능하고

두 가지 이상의 테이블을 조합해 작성한 뷰를 갱신할 때는 어떤 테이블을 갱신하면 좋을지 몰라 할 수 없다.

뷰에서 원래 테이블의 일부 열만 선택되었다면 데이터를 삽입하려고 해도 선택된 열 이외의 열에 기본값도 없고 NULL도 허용되지 않는 상황에서는 해당 열에 넣을 수 있는 값이 없어서 실질적으로 뷰로의 삽입이 불가능.

## 4. 서브쿼리의 실행

SELECT 문의 결과를 마치 데이터처럼 다루거나 수치처럼 취급해서 조건문에 이용하는 경우, 이런 쿼리를 서브쿼리라고 부른다.

### 5-1. 내부 결합

결합은 두 개의 테이블에서 필요한 열을 가지고 올 때 행을 결합하기 위한 조건을 "ON"으로 지정
내부결합에서는 이 ON으로 지정한 결합 조건에 일치하는 행만을 두 개의 테이블로부터 가져올 수 있다.
여기서는 두 개의 테이블에서 결합 조건에 일치하는 것만을 얻어낸다.

### 5-1. 외부 결합
어떤 테이블을 기준으로 잡고 그 테이블에 있는 정보를 모두 표현하되, 조건과 일치하는 다른 테이블의 값도 함께 표현하는 것을 의미한다.

**Comment by rawfishthelgh** [comment]
# 6장

- select 문의 기본은 “어디서 무엇을 가지고 온다”를 나타내는 것이다
- db에서 열의 수는 고정이지만, 행의 수는 무한정 늘어날 수 있기에 where 구문을 활용해 효율성 있는 데이터를 선별해 가져온다
- 연산자에는 우선순위가 존재한다. 하지만 우선순위를 확실히 하기 위해 괄호를 사용할 수 있다.

### sql 기초 기술 규칙

- sql문의 마지막에 딜리미터(;)를 붙인다
- 키워드의 대소문자를 구별하지 않는다
- 정수는 기존 그대로 쓴다. 문자열이나 날짜 시각은 작은따옴표(’’)로 감싼다
- 단어는 반각 스페이스(한칸)나 개행으로 구분한다. 전각 스페이스(두칸)는 사용하지 않는다
- 중복된 값을 피하기 위해서는 distinct 키워드를 사용한다

### 검색결과 정렬

- 행의 순서는 “ORDER BY”란 명령으로 지정한다.
- ASC는 오름차순 DESC는 내림차순 정렬이다.

### 테이블을 요약하는 함수

- 데이터에 대해 어떠한 집계를 수행하는 함수를 집계함수라 부른다. 종류는 아래와 같다
- count : 테이블의 행 수
- sum : 테이블의 수치 데이터 합계
- avg : 테이블의 수치 데이터 평균
- max : 테이블의 임의열 데이터 중 최대값
- min : 테이블의 임의열 데이터 중 최소값

### 문자열 집약

- sql 표준에는 없으나, mysql에는 문자열을 집약하는 GROUP_CONCAT 함수가 존재한다
- 문자열에 대한 집계를 문자열의 결합으로 수행한다. 콤마로 구분되는 매우 긴 데이터를 결과로 돌려 준다.
- distinct를 집약함수로 이용한다면 group_concat에서 반환되는 문자열에서 중복을 제거할 수 있다

`select group_concat(district) from city where countrycode = 'KOR'` 

→ Seoul,Seoul,Pusan

`select group_concat(DISTINCT district) from city where countrycode = 'KOR'`

→ Seoul,Pusan

### 데이터를 그룹으로 나누는 Group By

- group by를 사용해 대상이 되는 데이터를 몇 개의 그룹으로 나눠서 집약할 수 있다. group by에 지정한 열을 하나의 그룹으로 각 그룹의 결과를 보여준다

### 나눈 그룹에 조건 추가

- 그룹으로 나눠서 집약한 값(`count(*`)에 대해 조건을 설정할 때 where을 사용할 수 없다. count 같은 집약함수는 오직 select와 order by와 having에서만 사용할 수 있다
- 그룹마다 집약한 값을 조건으로 선택하려면 having 뒤에 조건을 지정한다

`select discrict, count(*) from city where countrycode ='KOR' group by district having count(**) = 6;`

`order by count(*) desc;`

- 작성순서는 반드시 아래를 지켜야 한다
    1. select
    2. from
    3. where
    4. group by
    5. having
    6. order by

### 데이터를 변경하는 update

`update 테이블명 set 열명 = 값 where 조건;`

위 형태로 where 로 갱신할 대상을 간추려 set에 해당하는 데이터를 갱신한다

`set 열명1 =값1, 열명2=값2` 와 같은 구문을 사용하면 복수 개의 열을 동시에 갱신할 수 있다

### INSERT

insert 문의 기본 구문은 아래와 같다

`insert into 테이블명(열1,열2...) values (값1,값2...);`

- 테이블명 뒤의 열 리스트와 values 뒤의 값 리스트는 수와 데이터형이 일치해야 한다
- mysql에는 insert문 하나로 신규 복수 행을 입력할수 있는 기능이 있다

`insert into city(name,district) values ('gimpo','kor'), ('seoul','kor')`

두 개의 insert 쿼리를 1번으로 정리해 처리 시간이 짧아진다

### DELETE

아래와 같은 형태로 데이터를 제거한다

`delete from 테이블명 where 조건;`

## 뷰를 작성하고 복수 테이블에서 선택하기

### 뷰를 사용하는 이점

- 뷰는 테이블과 동일하지만 데이터는 갖고 있지 않으며, 테이블에 대한 select를 가지고 있다. 이점은 아래와 같다
    1. 복잡한 select문을 일일이 매번 기술할 필요가 없다
    2. 필요한 열과 행만 사용자에게 보여줄 수 있꼬, 갱신 시에도 뷰 정의에 따른 갱신으로 한정할 수 있다
    3. 1과 2의 이점을 데이터 저장 없이(기억장치의 용량을 사용하지 않고) 실현할 수 있다. 또한 뷰를 제거해도 참조하는 테이블은 영향을 받지 않는다.

### 뷰 작성

`create view 뷰 명(열명1,열명2) as select문;`

- 뷰를 작성할 때는 위와 같이 create view 문을 사용한다

### 뷰로의 입력, 갱신의 제한

뷰로는 어떤 행이 대응하는지 모르거나, 어떤 값을 넣으면 좋을지 모르는 경우에는 갱신할 수 없다. group by나 distinct로 얻은 값을 갱신하는 경우는 결과의 기반이 되는 테이블의 n행 중 어떤 수치를 갱신하면 좋을지 모른다. 혹은 두 가지 이상의 테이블을 조합해 작성한 뷰를 갱신할 때 어떤 테이블을 선택해 갱신해야할지 알 수 없다. 실질적으로 뷰로의 삽입은 불가능하다.

### 서브쿼리

- select문의 결과가 단일 값인 경우, 이를 조건문에 이용할 수 있다. 이런 쿼리를 “서브쿼리”라고 부른다.

### 결합

- sql은 단일 테이블 뿐 아니라 2개 이상의 테이블을 대상으로 실행할 수 있다. 이때 결합(JOIN)을 이용한다.
- 2개 테이블의 열을 결합시키기 위한 ‘결합조건’을 지정한다

### 내부결합

- 행을 결합하기 위한 조건을 “ON”으로 지정하면 내부결합을 수행할 수 있다.
- 내부결합은 둘 이상의 테이블에 해당하는 공통 조건만을 가져와 새 테이블로 만든다(교집합)

### 외부결합

- 조건에 일치하는 값만을 가져오는게 아닌, 한 쪽 테이블을 기준으로 전체행을 표시하고, 반대쪽은 결합 조건에 맞는 값이 있으면 표시하며 테이블을 만든다
- 결합 조건을 `left outer join` 을 사용하여, 왼쪽 테이블의 전체 행이 표시되고, 다른 테이블의 행 데이터는 결합 조건에 맞으면 그 데이터를 넣고, 일치하지 않으면 ‘null’을 넣는다.
- 반대로는 `right outer join` 을 사용 가능하다. 그러나 보통 left가 더 직관적이고 쉬워 많이 사용한다

**Comment by seokhwan-an** [comment]
<h1>6장 SQL 문의 기본</h1>
<h3>💡목표</h3>
<p>SQL문 중 SELECT문 (조회)을 통해서 SQL에 익숙해지자</p>
<h3>SELECT 문으로 테이블 내용을 살펴보자</h3>
<ol>
<li>테이블을 지정해서 전체 데이터 조회하기 <code>city</code></li>
</ol>
<pre><code class="language-sql">SELECT * FROM city;
</code></pre>
<ol>
<li>조건을 추가해서 데이터 조회하기 <code>한국의 도시만</code></li>
</ol>
<pre><code class="language-sql">SELECT * FROM city WHERE countrycode = 'KOR';
</code></pre>
<ol>
<li>조건을 추가해서 데이터 조회하기 <code>한국의 도시에서 전라남도에 속하는 도시만</code></li>
</ol>
<pre><code class="language-sql">SELECT * FROM city WHERE district = 'chollanam';
</code></pre>
<ol>
<li>불필요한 열을 제거하고 데이터 조회하기 전체 <code>column이 아닌 도시명과 인구만</code></li>
</ol>
<pre><code class="language-sql">SELECT Name,Population FROM city WHERE district = 'chollanam';
</code></pre>
<ol>
<li>다양한 조건 추가하기 <code>전라남도의 도시명과 인구수가 15만명 보다 큰 경우만</code></li>
</ol>
<pre><code class="language-sql">SELECT Name,Population FROM city WHERE district = 'chollanam' AND Population &gt; 150,000;
</code></pre>
<ol>
<li>다양한 조건 추가하기 <code>중복되는 행 제거하기</code></li>
</ol>
<pre><code class="language-sql">SELECT DISTINCT district FROM city WHERE countrycode = 'KOR';
</code></pre>
<h3>테이블을 조작하는 SQL</h3>
<p><code>select * from city;</code> 는 디폴트 데이터베이스의 city를 SELECT한다.</p>
<p><code>select * from world.cit[y;](&lt;http://world.city&gt;)</code> 는 ‘데이터베이스명.테이블명’ 형태로 명시적으로 world 데이터베이스의 city를 SELECT한다.</p>
<h3>WHERE 조건이 필요한 이유</h3>
<ul>
<li>
<p>테이블의 명은 초기에 설정하면 명시적으로 늘리지 않는 한 고정이다.</p>
</li>
<li>
<p>행의 경우 제한 없이 늘릴 수 있는데 필요한 데이터를 효율적으로 가져오기 위해서는 SELECT 해오는 행수를 줄여야한다.</p>
</li>
<li>
<p>그래서 Where절을 이용해야 한다.</p>
</li>
<li>
<p>MySql 연산자 우선 순위 (위로 갈 수록 우선순위가 높다)</p>


연산자
--
INTERVAL
BINARY, COLLATE
-(단항 연감산), ~(단항 비트 반전)
^
*, /, DIV, %, MOD
-, +
<<, >>
&
 
=, < = >, ≥, ≤. <, <>, ≠, IS, LIKE, REGEXP, IN
BETWEEN, CASE, WHEN, THEN, ELSE
NOT
&&, AND
XOR
 
=(대입 등호), :=


</li>
</ul>
<h3>SQL의 기초적인 기술 규칙</h3>
<ul>
<li>SQL 문의 마지막에 딜리미터를 붙인다. (대부분은 세미콜론(;)이다.)</li>
<li>키워드의 대문자 소문자는 구별하지 않는다.</li>
<li>정수는 기존 그대로 쓴다. 문자열이나 날짜 시각은 작은땀옴표(’’)로 감싼다. 즉 숫자는 100으로 문자열은 ‘100’으로 지정한다.</li>
<li>단어는 반각 스페이스나 개행으로 구별한다.</li>
</ul>
<blockquote>
<p>💡반각 스페이스 VS 전각 스페이스</p>
</blockquote>
<h3>DISTINCT에 의한 중복 배제</h3>
<ul>
<li>선택한 행에서 중복된 값이 존재하는 상황에서 이를 없애고 싶을 때 ‘DISTINCT’ 키워드를 지정한다.</li>
<li>where는 조건으로 행을 추리는 것이고 distinct는 중복된 열의 값을 제거해 행을 추리는 방식이다.</li>
</ul>
<h3>SELECT 문을 응용해보자</h3>
<ol>
<li>검색결과를 정렬해보자 <code>한국 도시를 인구가 적은 순서대로 정렬하기</code></li>
</ol>
<pre><code class="language-sql">SELECT * FROM city WHERE countrycode = 'KOR' order by population;
</code></pre>
<ol>
<li>테이블을 집약해 보자 <code>조회한 데이터의 개수 구하기</code></li>
</ol>
<pre><code class="language-sql">SELECT COUNT(*) FROM city WHERE countrycode = 'KOR';
</code></pre>
<ol>
<li>테이블을 집약해 보자 <code>인구수의 최소값, 최대 값, 전체 값, 평균값 구하기</code></li>
</ol>
<pre><code class="language-sql">SELECT MIN(population), MAX(population), SUM(population), AVG(population) FROM city WHERE countrycode = 'KOR';
</code></pre>
<ol>
<li>조회 결과를 한 향으로 집약하기</li>
</ol>
<pre><code class="language-sql">SELECT GROUP_CONCAT(name) FROM city WHERE countrycode = 'KOR' and district = 'Chollabuk';
</code></pre>
<ol>
<li>테이블을 그룹으로 구분하기</li>
</ol>
<pre><code class="language-sql">SELECT district, count(*) FROM city WHERE countrycode = 'KOR' group by district;
</code></pre>
<h3>검색결과 정렬</h3>
<ul>
<li>검색 결과의 순서를 지정하지 않으면 테이블에 저장된 순서로 조회된다.</li>
<li>순서를 설정하기 위해서는 order by를 이용하면된다. (오름차순 정렬)</li>
<li>내림차순 정렬은 추가로 DESC 키워드를 넣으면 된다.</li>
<li>정렬을 수행할 때 행의 순서를 정확하게 하기 위해서는 정렬하려는 행은 한가지의 의미만을 가져야한다. (여러개의 행으로도 한가지의 의미를 만들 수 있다.)</li>
</ul>
<h3>테이블을 요약하는 함수</h3>
<ul>
<li>SQL에서 데이터에 대해 어떠한 조작이나 계산을 수행하는데 ‘함수’라는 도구를 시용한다.</li>
<li>함수는 2가지의 종류가 있다.
<ul>
<li>복수 행(이나 행의 값)에 대해 집계를 수행하는 함수 → count, min, max, sum, avg</li>
<li>집계 함수의 경우 기본적으로 NULL을 제외하고 집계를 한다. 단, COUNT(*)로 표기하여 NULL을 포함한 전체 행을 집계한다.</li>
</ul>
</li>
</ul>
<h3>문자열을 집약하는 GROUP_CONCAT</h3>
<ul>
<li>SUM은 ‘수치’에 대한 집계를 한다면 GROUP_CONCAT은 ‘문자열’에 대한 집계를 ‘문자열의 결합’으로 나타낸다.</li>
<li>‘문자열의 결합’은 콤마(,)로 구분한다.</li>
<li>‘문자열의 결합’으로 결과를 반환하기에 예상치 못한 길이의 결과를 반환할 수 있다.</li>
</ul>
<h3>DISTINCT로 중복 회피</h3>
<ul>
<li>중복되는 결과값을 제거하는 키워드이다.</li>
</ul>
<h3>데이터를 그룹으로 나누는 GROUP BY</h3>
<ul>
<li>대상이 되는 데이터를 몇 개의 그룹으로 나누어서 집약하는 키워드이다.</li>
<li>그룹을 나눌 때에는 키가 되는 열을 지정한다.</li>
<li>GROUP BY의 결과가 없으면 대상 데이터 전체로 집약해 1건의 결과만 돌려주지만, GROUP BY의 결과가 있으면 지정한 그룹 열마다 집약해서 그룹 열로 나눈 그룹 수의 결과를 되돌려준다.</li>
<li>GROUP BY에 추가적인 조건을 주기 위해서는 WHERE이 아닌 HAVING을 이용해야한다.</li>
<li>SELECT → FROM → WHERE → GROUP BY → HAVING → ORDER BY 순으로 SQL을 작성해야한다.</li>
</ul>
<h3>데이터를 갱신, 삽입, 제거해보자</h3>
<ol>
<li>데이터 갱신하기</li>
</ol>
<pre><code class="language-sql">UPDATE city SET name = 'Siheung' WHERE countrycode = 'KOR' and district = 'Kyonggi' and name = 'Shihung';
</code></pre>
<ol>
<li>데이터 삽입하기</li>
</ol>
<pre><code class="language-sql">INSERT INTO city VALUES (DEFAULT, 'Gimpo', 'KOR', 'Kyonggi', 359584);
</code></pre>
<ol>
<li>데이터 제거하기</li>
</ol>
<pre><code class="language-sql">DELETE FROM city WHERE id = 4000;
</code></pre>
<h3>데이터를 변경하는 UPDATE 문</h3>
<ul>
<li>SET으로 지정한 열 이외의 것은 변경되지 않는다.</li>
<li>주로 테이블 전체 대상으로 하는 것이 아닌 하나의 행을 기준으로 한다.</li>
<li>하나의 열만 수정가능한 것이 아닌 복수개의 열을 수정할 수 있다.</li>
</ul>
<h3>데이터를 입력하는 INSERT문</h3>
<ul>
<li>INSERT는 행의 단위로 수행되므로 행을 구성하는 열의 정의 정보를 포함한 ‘테이블 정의’를 정확하기 알아야한다.</li>
<li><code>show create table 테이블명\\G</code>를 통해 table의 DDL을 알 수 있다.</li>
<li>또는 <code>desc 테이블명</code> 을 통해 표의 형태로 한눈에 보기 쉽게 테이블에 정보를 알 수 있다.</li>
<li>INSERT 문의 기분 구문 : <code>INSERT INTO 테이블명(열1[, 열2, …]) VALUES (값1[, 값2, …]);</code></li>
<li>테이블명 뒤의 열의 개수와 VALUES 뒤의 값의 개수가 일치해야한다.</li>
<li>테이블에 정의된 모든 열에 대해 VALUES에서 값을 설정하면 테이블명의 열 리스트를 생략할 수 있다.</li>
<li>DEFAULT 키워드를 명시적으로 이용하는 것은 데이터를 INSERT 시 해당 기본값을 사용한다.라는 것을 의미한다.</li>
<li>ID에서는 주로 DEFAULT를 생략하여 암묵적으로 처리하는 경우도 있다.</li>
<li>데이터 입력에 자주 사용되는 구문으로는 <code>INSERT INTO 테이블1 SELECT * FROM 테이블2</code> 가 있다.
<ul>
<li>VALUES 대신에 테이블2에서 조회한 데이터를 테이블1에 삽입하는 경우다.</li>
</ul>
</li>
<li>여러 행을 한번에 삽입하는 경우에는 각각 INSERT를 처리하는 것 보다는 하나의 INSERT에서 VALUES 뒤에 괄호를 콤마(,)로 연결하는 것이 성능이 좋다.</li>
</ul>
<h3>데이터를 제거하는 DELETE 문</h3>
<ul>
<li>DELETE문의 기본 구문은 <code>DELETE FROM 테이블명</code> 이나 이는 지정한 테이블의 전체 행이 제거되므로 거의 사용하지 않고 갱신하는 대상을 WHERE로 추려낸 <code>DELETE FROM 테이블명 WHERE</code> 를 이용한다.</li>
</ul>
<h3>뷰를 작성하고 복수 테이블에서 선택해보자</h3>
<ol>
<li>뷰를 작성해 보자</li>
</ol>
<pre><code class="language-sql">CREATE VIEW citykyonggi AS SELECT id, name, population FROM city WHERE countrycode = 'KOR' AND district = 'Kyonggi';
</code></pre>
<ol>
<li>제약 조건이 추가된 뷰를 만들어보자</li>
</ol>
<pre><code class="language-sql">CREATE VIEW largecity AS SELECT * FROM city WHERE population &gt; 7000000 WITH CHECK OPTION;
</code></pre>
<ol>
<li>서브쿼리를 실행해보자</li>
</ol>
<pre><code class="language-sql"># 한국의 도시 중에서 인구가 평균 이상인 도시 수 구하기
SELECT COUNT(*) FROM cityKorea WHERE population &gt; (SELECT AVG(population) FROM cityKorea);
</code></pre>
<ol>
<li>결합(Inner Join, Outer Join)을 실행해보자</li>
</ol>
<pre><code class="language-sql">SELECT countrylanguage.*, country.name FROM countrylanguage INNER JOIN country ON countrylanguage.countrycode = country.code WHERE language = 'Korean';
</code></pre>
<h3>뷰</h3>
<ul>
<li>뷰를 사용하는 이점
<ul>
<li>복잡한 SELECT문을 일일이 매번 기술할 필요가 없다.</li>
<li>필요한 열과 행만 사용자에게 보여줄 수 있고, 갱신 시에도 뷰 정의에 따른 갱신으로 한정할 수 있다.</li>
<li>위의 두 이점을 데이터 저장 없이 실현할 수 있다. 또한, 뷰를 제거도 참조하는 테이블은 영향을 받지 않는다.</li>
</ul>
</li>
<li>뷰는 <code>CREATE VIEW</code>를 통해 생상된다.</li>
<li>뷰의 경우 입력 및 갱신에 있어서 제한이 존재한다.
<ul>
<li>기본적으로 어떤 행이 대응되는지 모르거나 어떤 값을 넣으면 좋을지 모르는 경우에는 갱신이 불가능하다. (ex GROUP BY로 집약한 수치나 distinct로 어떤 값, 2가지 이상의 테이블의 조합으로 이루어진 뷰를 갱신하는 경우)</li>
</ul>
</li>
</ul>
<h3>서브쿼리</h3>
<ul>
<li>단일값으로 테이블이 구성되는 경우 이를 SELECT로 조회한 결과를 마치 데이터나 수치처럼 취급해 조건문에 활용할 수 있는데 이를 서브쿼리라고 한다.</li>
</ul>
<h3>조인</h3>
<ul>
<li>2개 이상의 테이블을 대상으로 데이터를 조회해야하는 경우가 있는데 이 때 조인을 이용한다.</li>
<li>내부조인
<ul>
<li>2개의 테이블에서 필용한 열을 가지고 올 때 행을 결합하기 위한 조건을 <code>ON</code>을 지정한다.</li>
<li>결합한 결과에 대해서 WHERE, ORDER BY, LIMIT이나 GROUP BY, HAVING을 지정할 수 있다.</li>
</ul>
</li>
<li>외부조인
<ul>
<li>한 쪽 테이블을 기준으로 행의 수를 고정해 한눈에 2개의 테이블의 데이터를 조회하는 방식이다.</li>
<li>같은 열이 존재하지 않는 경우에는 null 값이 들어가게 된다ㅣ.</li>
<li>기준이 되는 테이블에 따라 left 외부조인과 right 외부조인이 존재한다. 주로 left 외부조인을 이용한다.</li>
</ul>
</li>
</ul>

**Comment by somefood** [comment]
# SQL문의 기본

## SQL의 기초적인 기술 규칙

- SQL 문의 마지막에 딜리미터(Delimiter)를 붙여줌. 일반적으로 `;`
- 키워드(예약어)는 대소문자를 구분하지 않음
- 문자열과 날짜 시각은 작은 따옴표를('') 감쌈 큰 따옴표는 안 됨
- 단어는 스페이스나 개행으로 구별 됨

### DISTINCT 키워드

- 중복을 제거할 때 사용되는 키워드
- SELECT DISTINCT district FROM city WHERE countrycode = 'KOR';

### 집계 함수

- COUNT: 테이블의 행수를 알려주는 함수
- SUM: 테이블의 수치 데이터를 합계하는 함수
- AVG: 테이블의 수치 데이터 평균을 구하는 함수
- MAX: 테이블의 임의열 데이터 중 최대값을 구하는 함수
- MIN: 테이블의 임의별 데이터 중 최소값을 구하는 함수

### GROUP BY

- 데이터를 그룹으로 모을 때 사용
- 위의 집계 함수를 활용할 때 사용됨
- EX) 나라별 인구 수 보여줄 때
- `SELECT count(*) FROM world GROUP BY country;`

### HAVING BY

- 나눈 그룹에서 조건을 추가할 때 사용
- EX) 인구가 5000이 넘는 나라 목록 보여줄 때
- `SELECT * FROM world GROUP BY country HAVING population > 5000;`

## 데이터 삽입

- INSERT 테이블() VALUE(): 단건으로 넣을 때
- INSERT 테이블() VALUES(): 여러건 넣을 때

> show create table 테이블명\G <- 테이블 구조 깔끔하게 출력

## 데이터 수정

- UPDATE 테이블 SET 컬럼=값 WHERE [조건]
- **WHERE절 없이 입력하면 전체 행에 영향가니 매우 주의!!**

## 데이터 삭제

- DELETE FROM 테이블 WHERE [조건]
- **WHERE절 없이 입력하면 전체 행에 영향가니 매우 주의!!**

## 뷰(VIEW)

- 테이블에 대한 SELECT 결과를 가상의 테이블로 가지고 있음
  - 복잡한 SELECT 문을 일일이 매번 기술할 필요 없음
  - 필요한 열과 행만 사용자에게 보여줄 수 있고, 갱신 시에도 뷰 정의에 따른 갱신으로 한정할 수 있음
  - 기억장치의 용량을 사용하지 않아 데이터 용량을 절감할 수 있음
- CREATE VIEW citykyonggi AS SELECT ID, NAME, POPULATION FROM city WHERE countrycode = 'KOR' AND district='Kyonggi';

## 서브쿼리(SubQuery)

- SELECT 문의 결과를 데이터처럼 다루거나 수치처럼 취급해 조건문에 이용할 수 있는 구절
- `SELECT count(*) FROM citykorea where population > (SELECT AVG(population) FROM citykorea);`
